### PR TITLE
Add basic tests for GetMajor/Minor HTTP versions in the response.

### PR DESCRIPTION
**Pull request by ahsonkhan**

### CODE REVIEW

File: sdk/core/azure-core/test/ut/http_test.cpp [hunk 1]
```diff
@@ -121,6 +121,24 @@ namespace Azure { namespace Core { namespace Test {
         (std::pair<std::string, std::string>("valid3", "header3")));
   }
 
+  TEST(TestHttp, RawResponseHttpVersions)
+  {
+    Http::RawResponse responseHttp10(1, 0, Http::HttpStatusCode::Accepted, "Test");
```
Comment (RickWinter) [inline sdk/core/azure-core/test/ut/http_test.cpp:126]: We should parameterize this test so that we can test all scenarios, positive cases and negative cases.

**Review by RickWinter — COMMENTED** [review]
Please add tests Min/Max, i.e. 0.9, 9999999999999999999999.999999999999999999, -1.0, etc...

File: sdk/core/azure-core/test/ut/http_test.cpp (hunk 1 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/http_test.cpp:126]: We have no negative case validation in place.

What do you expect as the behavior in those cases?

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/http_test.cpp:126]: I would expect invalid values to throw in the Raw response ctor. Until then, I don't want to codify in tests the current behavior for such inputs.

File: sdk/core/azure-core/test/ut/http_test.cpp [hunk 2]
```diff
@@ -121,6 +121,32 @@ namespace Azure { namespace Core { namespace Test {
...
+    std::pair<std::int32_t, std::int32_t> httpVersion;
+  };
+
+  INSTANTIATE_TEST_SUITE_P(
+      TestHttp,
+      ParameterizedTestForHttpVersions,
+      ::testing::Values(
+          std::make_pair(0, 9),
+          std::make_pair(1, 0),
+          std::make_pair(1, 1),
```
Comment (LarryOsterman) [inline sdk/core/azure-core/test/ut/http_test.cpp:136]: I'd also add 1.5 to the list, but otherwise it looks fine.

Have you verified that this changes code coverage numbers?

**Review by LarryOsterman — APPROVED** [review]

File: sdk/core/azure-core/test/ut/http_test.cpp [hunk 3]
```diff
@@ -121,6 +121,32 @@ namespace Azure { namespace Core { namespace Test {
...
+          std::make_pair(2, 0)));
```
Comment (RickWinter) [inline sdk/core/azure-core/test/ut/http_test.cpp:137]: Need to validate the edge cases as well.
https://datatracker.ietf.org/doc/html/rfc9110#name-protocol-version 

I would expect the rawresponse parser to throw a 505 if the httpversion is < 0.0. I would expect throws for >3.x.
https://datatracker.ietf.org/doc/html/rfc9110#name-505-http-version-not-suppor

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/http_test.cpp:137]: That's the responsibility of the parser logic/transport layer. Nothing in the RawResponse type/ctor itself does any validation or checks for such behavior. This unit test is purely about the RawResponse type.

I agree, we should add those edge case tests as part of receiving such responses when going through the transport parsers/http pipeline.

In the Raw response ctor itself, we decided not to do argument validation for negative values. Currently, the stance is "garbage in/garbage out".

Comment (RickWinter) [reply sdk/core/azure-core/test/ut/http_test.cpp:137]: In that case I'd add int32_t Min/Max combinations. They aren't legal http but they are legal ints

File: sdk/core/azure-core/test/ut/http_test.cpp (hunk 2 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/http_test.cpp:136]: I checked last week, and the two method calls were covered, indirectly, from one of the curl test that exercised the method calls (as part of keep alive). Therefore, it doesn't change the coverage numbers. Maybe the function call is easily inlined given how small it is.

You can see the yellow/partial cover on line 957, here:
https://github.com/Azure/azure-sdk-for-cpp/issues/5612

File: sdk/core/azure-core/test/ut/http_test.cpp (hunk 3 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/http_test.cpp:137]: I can add int max, but int min is a negative value. I don't think we should be codify in the behavior of supporting negative values here, passed in to RawResponse, because that's an unsupported/undefined scenario.

I don't want a unit test to fail, if we were to add a check in the RawResponse ctor that throws on negative values in the future (which is a reasonable thing to do). With such a change, existing tests shouldn't fail.

Adding a test for the parser where negative values (or other malformed/unsupported values) showing up in the response header, makes sense. So let's add those, as part of future test enhancements, such as:
#5612

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/http_test.cpp:137]: Reading the doc you linked, returning 505 as a status code, is the **server's** responsibility. It isn't the responsibility of the client/SDK to validate the HTTP version in the response for not supported HTTP versions.

If a service call returns a 505 with the response, we'd treat that as any other non-success status code and retry or throw as we'd normally do, based on our existing 4xx and 5xx status code handling logic.

Comment (LarryOsterman) [reply sdk/core/azure-core/test/ut/http_test.cpp:137]: I'm with Ahson here, for two reasons:

1) It's not responsibility to invent HTTP responses. We should faithfully respect the server response if possible.
2) The only time we should fail on a server response is if it is impossible to properly interpret the response.

That last one is because the Azure SDK has a blanket policy of not validating server responses. Validation of server responses is treated similar to how customer provided inputs are validated - validation should be done only to ensure correct functioning of the client - so a null pointer check which turns a SEGFAULT exception into a C++ exception that can be handled makes sense. 

Similarly, it is reasonable for the C++ libCURL transport to throw an exception if an HTTP response cannot be parsed correctly or if the response version is unrecognizable (something other than 1.x for instance). On the other hand, IMHO it is *not* reasonable for the WinHTTP transport to throw an exception on a non 1.x HTTP version because WinHTTP supports HTTP/2.

Comment (RickWinter) [reply sdk/core/azure-core/test/ut/http_test.cpp:137]: > Reading the doc you linked, returning 505 as a status code, is the **server's** responsibility. It isn't the responsibility of the client/SDK to validate the HTTP version in the response for not supported HTTP versions.
> 
> If a service call returns a 505 with the response, we'd treat that as any other non-success status code and retry or throw as we'd normally do, based on our existing 4xx and 5xx status code handling logic.
> 
> > The 505 (HTTP Version Not Supported) status code indicates that the server does not support, or refuses to support, the major version of HTTP that was used in the request message.
> 
> Are you recommending, we concoct our own http status code 505, if the server doesn't return that, if the response happen to contain an HTTP version < 0.0 or above 3.x? That seems a bit odd to me, for us to do as an SDK. We could throw a transport exception for malformed responses, but I wouldn't change the returned HTTP status code, or make one up.

I agree, we shouldn't concoct anything. Clients don't validate the request. That is the job of the transports to only send appropriate headers/bodies/... per the http version specs. On return, I'd expect the parser(s) to parse out any int value and return it. Anything that isn't an int should fail. Again that would be in the parser logic.
This API I expect to return any valid int32_t value (whether or not that value is legal per http spec is not up to this API). Thus the edge cases for this api are int32_t min and int32_t max.

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/http_test.cpp:137]: > validation should be done only to ensure correct functioning of the client - so a null pointer check which turns a SEGFAULT exception into a C++ exception that can be handled makes sense.

> That is the job of the transports to only send appropriate headers/bodies/... per the http version specs. On return, I'd expect the parser(s) to parse out any int value and return it. Anything that isn't an int should fail.

Regarding that aspect (which is out of scope for this PR), I think it's valuable for us to fuzz and harden our transports. Here's a follow-up to tackle that, potentially for MQ:
https://github.com/Azure/azure-sdk-for-cpp/issues/5624

Comment (RickWinter) [reply sdk/core/azure-core/test/ut/http_test.cpp:137]: Why do not want to test 0.0 here or any other int32_t value? They are legal per the implementation of the current RawResponse class?

>I don't want a unit test to fail, if we were to add a check in the RawResponse ctor that throws on negative values in the future (which is a reasonable thing to do). With such a change, existing tests shouldn't fail.

I like tests to fail in that case. It lets me see where existing behavior changed. I might need to update a test to account for the behavior change, which is ok at times.

> I think this requires a follow-up discussion to decide what behavior of RawResponse we actually want to codify within the test (outside this PR thread).

I would consider this PR to be doing that. ;)

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/http_test.cpp:137]: > They are legal per the implementation of the current RawResponse class?

Ideally, tests shouldn't cater to the (current) implementation, but rather the other way around.

It is an open question whether how things are implemented is what we actually want. I expect any line of code in our SDK like the following to throw (if it was either written intentionally or accidentally) since it indicates a bug _somewhere_:

```C++
RawResponse resp(-1, -2, HttpStatusCode::Ok, "Test"); // invalid code, shouldn't succeed

// I don't want to codify in tests that the expected behavior here should be as follows, regardless of what the current implementation is lacking
cout << resp.GetMajorVersion(); // -1
```

> I like tests to fail in that case. It lets me see where existing behavior changed. I might need to update a test to account for the behavior change, which is ok at times.

I agree it is OK at times. I don't think this is one of those times. Tests failing should indicate some bug either in the product or in how the test is written/change in requirements OR some unintended design/behavior changing. This isn't that.

There is nothing preventing us to write such tests in the future, once we agree on the behavior we want.
